File: sdks/go/client_test.go
package stackless

import (
	"encoding/json"
	"errors"
	"os"
	"os/exec"
	"path/filepath"
	"slices"
	"testing"
)

func TestHostExecutionRequiresExplicitCallerFlag(t *testing.T) {
	for _, allowed := range []bool{false, true} {
		for _, request := range []UpRequest{
			UpCreate(Create{On: "local", AllowHostExecution: allowed}),
			UpResume(Resume{Name: "demo", AllowHostExecution: allowed}),
		} {
			args, err := upArgs(request)
			if err != nil {
				t.Fatal(err)
			}
			if slices.Contains(args, "--allow-host-execution") != allowed {
				t.Fatalf("host grant %v produced args %v", allowed, args)
			}
		}
	}
}

type fakeRunner struct {
	byKey map[string]map[string]any
}

func (f fakeRunner) Run(bin string, args []string, cwd string) ([]byte, []byte, int, error) {
	key := keyArgs(args[1:]...)
	payload := f.byKey[key]
	out, _ := json.Marshal(payload)
	code := 0
	if ok, _ := payload["ok"].(bool); !ok {
		code = 1
	}
	return out, nil, code, nil
}

func keyArgs(args ...string) string {
	b, _ := json.Marshal(args)
	return string(b)
}

func TestControllerIsForwardedToSubmissionsAndOperationReads(t *testing.T) {
	base := New("/fake/stackless", "")
	base.SetRunner(fakeRunner{byKey: map[string]map[string]any{
		keyArgs("--controller", "ssh://deploy@builder", "down", "demo", "--no-wait"):              {"ok": true, "operation": map[string]any{"id": "op"}},
		keyArgs("--controller", "ssh://deploy@builder", "operation", "get", "op", "--after", "0"): {"ok": true, "result": map[string]any{"operation": map[string]any{"id": "op"}, "events": []any{}}},
	}})
	c := base.WithController("ssh://deploy@builder")
	if base.controller != "" {
		t.Fatal("WithController mutated the original client")
	}
	if _, err := c.SubmitDown("demo"); err != nil {
		t.Fatal(err)
	}
	if _, err := c.Operation("op", 0); err != nil {
		t.Fatal(err)
	}
}

func TestUpOriginsAndIntegrations(t *testing.T) {
	payload := map[string]any{
		"schema_version": 1,
		"ok":             true,
		"instance":       "demo",
		"instance_id":    "owner-1",
		"substrate":      "local",
		"executed":       []any{"start:web"},
		"skipped":        []any{},
		"duration_ms":    12,
		"steps":          []any{},
		"origins": []any{
			map[string]any{"service": "web", "origin": "http://demo.localhost:4444/"},
		},
		"placements": map[string]any{"workloads": map[string]any{"web": "local", "api": "fly"}, "resources": map[string]any{"clerk": "local"}},
		"endpoints":  map[string]any{"public": map[string]any{"workload": "web", "url": "https://public.example.test", "source": "declared"}},
		"integrations": map[string]any{
			"clerk": map[string]any{"secret_key": map[string]any{"kind": "secret_ref", "instance_id": "owner-1", "integration": "clerk", "output": "secret_key"}},
		},
	}
	c := New("/fake/stackless", "")
	c.SetRunner(fakeRunner{byKey: map[string]map[string]any{
		keyArgs("up", "--on", "local", "--name", "demo"): payload,
	}})

	out, err := c.Up(UpCreate(Create{On: "local", Name: "demo"}))
	if err != nil {
		t.Fatal(err)
	}
	if out.Placements.Workloads["api"] != "fly" || out.Placements.Resources["clerk"] != "local" {
		t.Fatal(out.Placements)
	}
	if out.Endpoints["public"].Source != "declared" || out.EndpointURLs()["public"] != "https://public.example.test" {
		t.Fatalf("endpoint missing: %v", out.Endpoints)
	}
	if out.Origins["web"] != "http://demo.localhost:4444/" {
		t.Fatalf("origin: %q", out.Origins["web"])
	}
	if out.Integrations["clerk"]["secret_key"].InstanceID != "owner-1" {
		t.Fatalf("integration key missing")
	}
}

func TestDownErrorCode(t *testing.T) {
	c := New("/fake/stackless", "")
	c.SetRunner(fakeRunner{byKey: map[string]map[string]any{
		keyArgs("down", "missing"): {
			"ok": false,
			"error": map[string]any{
				"code":    "instance_not_found",
				"message": "no such instance",
			},
		},
	}})
	_, err := c.Down("missing")
	var se *Error
	if !errors.As(err, &se) || se.Code != "instance_not_found" {
		t.Fatalf("expected instance_not_found, got %v", err)
	}
}

func TestListCheck(t *testing.T) {
	c := New("/fake/stackless", "")
	c.SetRunner(fakeRunner{byKey: map[string]map[string]any{
		keyArgs("list"): {
			"schema_version":      1,
			"ok":                  true,
			"instances":           []any{},
			"persistence_warning": "leases ephemeral",
		},
		keyArgs("check", "stackless.toml"): {
			"schema_version": 1,
			"ok":             true,
			"stack":          "demo",
			"services":       []any{"web"},
			"graph":          map[string]any{"nodes": []any{}},
		},
	}})
	listed, err := c.List()
	if err != nil || len(listed.Instances) != 0 {
		t.Fatalf("list: %v", err)
	}
	check, err := c.Check("stackless.toml", "")
	if err != nil || check.Stack != "demo" {
		t.Fatalf("check: %v", err)
	}
}

func TestDefaultRunnerDrainsStderrBeforeStdoutCloses(t *testing.T) {
	if _, err := exec.LookPath("python3"); err != nil {
		t.Skip("python3 required")
	}
	// Flood stderr past typical pipe capacity, then emit stdout JSON — the old
	// serial StdoutPipe/StderrPipe drain hung here.
	stdout, stderr, code, err := defaultRunner{}.Run("python3", []string{"-c", `
import sys
sys.stderr.write("x" * 256000)
sys.stderr.flush()
sys.stdout.write('{"ok": true, "schema_version": 1}')
sys.stdout.flush()
`}, "")
	if err != nil {
		t.Fatalf("run: %v", err)
	}
	if code != 0 {
		t.Fatalf("exit %d stderr=%q", code, stderr)
	}
	var data map[string]any
	if json.Unmarshal(stdout, &data) != nil || data["ok"] != true {
		t.Fatalf("stdout: %s", stdout)
	}
	if len(stderr) < 256000 {
		t.Fatalf("stderr truncated: %d", len(stderr))
	}
}

func TestRejectsPlaintextAndForeignReferences(t *testing.T) {
	for _, ref := range []any{"sk_test_CANARY", map[string]any{"kind": "secret_ref", "instance_id": "foreign", "integration": "clerk", "output": "secret_key"}} {
		_, err := secretRefs(map[string]any{"clerk": map[string]any{"secret_key": ref}}, "owner-1")
		if err == nil {
			t.Fatal("unsafe reference accepted")
		}
	}
}

func TestInvalidEndpointBindings(t *testing.T) {
	for _, raw := range []any{[]any{}, "url", map[string]any{"public": map[string]any{"workload": "web", "url": "https://example.test", "source": "invented"}}, map[string]any{"public": map[string]any{"url": "https://example.test", "source": "provider"}}} {
		if _, err := endpointBindings(raw); err == nil {
			t.Fatalf("accepted invalid endpoint: %v", raw)
		}
	}
}

func TestGeneratedEndpointBindingNames(t *testing.T) {
	generated, err := os.ReadFile("../../crates/stackless-idl/testdata/endpoints.go")
	if err != nil {
		t.Fatal(err)
	}
	root := t.TempDir()
	files := map[string][]byte{
		"go.mod":       []byte("module endpointfixture\n\ngo 1.22\n"),
		"endpoints.go": generated,
		"endpoints_test.go": []byte(`package stacklessbind
import "testing"
func TestBindings(t *testing.T) {
    if _, err := BindEndpoints(map[string]string{}); err == nil { t.Fatal("missing endpoint accepted") }
    bindings, err := BindEndpoints(map[string]string{"native-api":"http://native.example.test", "public-api":"https://public.example.test/v1"})
    if err != nil { t.Fatal(err) }
    if bindings.NativeApi != "http://native.example.test" || bindings.PublicApi != "https://public.example.test/v1" { t.Fatalf("wrong bindings: %+v", bindings) }
}
`),
	}
	for name, contents := range files {
		if err := os.WriteFile(filepath.Join(root, name), contents, 0600); err != nil {
			t.Fatal(err)
		}
	}
	cmd := exec.Command("go", "test", "./...")
	cmd.Dir = root
	if output, err := cmd.CombinedOutput(); err != nil {
		t.Fatalf("generated endpoint bindings failed: %v\n%s", err, output)
	}
}

func TestInvalidPlacements(t *testing.T) {
	for _, value := range []any{[]any{}, "local", map[string]any{}, map[string]any{"workloads": map[string]any{"api": false}, "resources": map[string]any{}}, map[string]any{"workloads": map[string]any{}, "resources": map[string]any{"db": ""}}} {
		payload := map[string]any{"ok": true, "instance": "demo", "instance_id": "owner-1", "substrate": "local", "origins": []any{}, "placements": value}
		c := New("/fake/stackless", "")
		c.SetRunner(fakeRunner{byKey: map[string]map[string]any{keyArgs("up", "--on", "local"): payload}})
		if _, err := c.Up(UpCreate(Create{On: "local"})); err == nil {
			t.Fatalf("accepted invalid placements: %#v", value)
		}
	}
}

File: sdks/typescript/src/client.test.ts
import { chmodSync, mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, describe, expect, it } from "vitest";

import { resolveStacklessBin } from "./bin.js";
import { Client, endpointUrls, StacklessError, type SpawnRunner } from "./client.js";
import { parseEnvelope } from "./envelope.js";

describe("parseEnvelope", () => {
  it("throws StacklessError with code on failure envelope", () => {
    const stdout = JSON.stringify({
      ok: false,
      error: { code: "instance_not_found", message: "missing demo" },
    });
    try {
      parseEnvelope(stdout);
      expect.unreachable("expected throw");
    } catch (err) {
      expect(err).toBeInstanceOf(StacklessError);
      expect((err as StacklessError).code).toBe("instance_not_found");
      expect((err as StacklessError).message).toBe("missing demo");
    }
  });
});

describe("Client", () => {
  const envBin = process.env.STACKLESS_BIN;

  it("forwards the controller to submissions and operation reads", async () => {
    const calls: string[][] = [];
    const run: SpawnRunner = (_bin, args) => {
      calls.push(args);
      return { status: 0, stderr: "", stdout: JSON.stringify({ ok: true, operation: { id: "op" }, result: { operation: { id: "op" }, events: [] } }) };
    };
    const client = new Client({ bin: "/fake/stackless", controller: "ssh://deploy@builder", run });
    await client.submitDown("demo");
    await client.operation("op");
    for (const args of calls) expect(args.slice(0, 3)).toEqual(["--json", "--controller", "ssh://deploy@builder"]);
  });

  it.each([false, true])("requires an explicit caller host grant: %s", async (allowed) => {
    const run: SpawnRunner = (_bin, args) => {
      expect(args.includes("--allow-host-execution")).toBe(allowed);
      return { status: 0, stderr: "", stdout: JSON.stringify({ ok: true, instance: "demo", instance_id: "owner-1", substrate: "local", origins: [] }) };
    };
    const client = new Client({ bin: "/fake/stackless", run });
    await client.up({ kind: "create", on: "local", allowHostExecution: allowed });
    await client.up({ kind: "resume", name: "demo", allowHostExecution: allowed });
  });

  afterEach(() => {
    if (envBin === undefined) {
      delete process.env.STACKLESS_BIN;
    } else {
      process.env.STACKLESS_BIN = envBin;
    }
  });

  it("maps up origins and integrations", async () => {
    const run: SpawnRunner = (_bin, args) => {
      expect(args[0]).toBe("--json");
      expect(args[1]).toBe("up");
      expect(args).toContain("--on");
      expect(args).toContain("local");
      return {
        status: 0,
        stderr: "",
        stdout: JSON.stringify({
          schema_version: 1,
          ok: true,
          instance: "demo",
          instance_id: "owner-1",
          substrate: "local",
          origins: [
            { service: "web", origin: "http://demo.localhost:4444/" },
            { service: "api", origin: "http://api.demo.localhost:4444/" },
          ],
          placements: { workloads: { web: "local", api: "fly" }, resources: { clerk: "local" } },
          endpoints: { public: { workload: "web", url: "https://public.example.test", source: "declared" } },
          integrations: {
            clerk: { secret_key: { kind: "secret_ref", instance_id: "owner-1", integration: "clerk", output: "secret_key" } },
          },
        }),
      };
    };

    const client = new Client({ bin: "/usr/bin/stackless", run });
    const outcome = await client.up({
      kind: "create",
      name: "demo",
      on: "local",
    });

    expect(outcome.instance).toBe("demo");
    expect(outcome.endpoints.public.source).toBe("declared");
    expect(endpointUrls(outcome)).toEqual({ public: "https://public.example.test" });
    expect(outcome.substrate).toBe("local");
    expect(outcome.placements.workloads.api).toBe("fly");
    expect(outcome.placements.resources.clerk).toBe("local");
    expect(outcome.origins).toEqual({
      web: "http://demo.localhost:4444/",
      api: "http://api.demo.localhost:4444/",
    });
    expect(outcome.integrations.clerk.secret_key).toEqual({ kind: "secret_ref", instance_id: "owner-1", integration: "clerk", output: "secret_key" });
  });

  it("defaults integrations to {}", async () => {
    const run: SpawnRunner = () => ({
      status: 0,
      stderr: "",
      stdout: JSON.stringify({
        ok: true,
        instance: "x",
        instance_id: "owner-1",
        substrate: "local",
        origins: [],
      }),
    });
    const client = new Client({ run });
    const outcome = await client.up({ kind: "create", on: "local" });
    expect(outcome.integrations).toEqual({});
  });

  it("propagates CLI error envelopes from up", async () => {
    const run: SpawnRunner = () => ({
      status: 1,
      stderr: "boom",
      stdout: JSON.stringify({
        ok: false,
        error: { code: "bad_argument", message: "invalid lease" },
      }),
    });
    const client = new Client({ run });
    await expect(
      client.up({ kind: "create", on: "local" }),
    ).rejects.toMatchObject({
      code: "bad_argument",
      message: "invalid lease",
    });
  });
});

describe("resolveStacklessBin", () => {
  const previous = process.env.STACKLESS_BIN;

  afterEach(() => {
    if (previous === undefined) {
      delete process.env.STACKLESS_BIN;
    } else {
      process.env.STACKLESS_BIN = previous;
    }
  });

  it("prefers STACKLESS_BIN over default", () => {
    process.env.STACKLESS_BIN = "/custom/stackless";
    expect(resolveStacklessBin()).toBe("/custom/stackless");
    expect(new Client().resolvedBin()).toBe("/custom/stackless");
  });

  it("prefers explicit bin over STACKLESS_BIN", () => {
    process.env.STACKLESS_BIN = "/from-env";
    expect(resolveStacklessBin("/explicit")).toBe("/explicit");
    expect(new Client({ bin: "/explicit" }).resolvedBin()).toBe("/explicit");
  });
});


describe("async process transport", () => {
  it("allows timers to run while the CLI is executing", async () => {
    const dir = mkdtempSync(join(tmpdir(), "stackless-async-"));
    const bin = join(dir, "stackless");
    writeFileSync(bin, `#!${process.execPath}\nsetTimeout(() => console.log(JSON.stringify({ok:true})), 100);\n`);
    chmodSync(bin, 0o755);
    try {
      let timerFired = false;
      setTimeout(() => { timerFired = true; }, 0);
      const waiting = new Client({ bin }).check("unused.toml");
      await waiting;
      expect(timerFired).toBe(true);
    } finally { rmSync(dir, { recursive: true, force: true }); }
  });
});

for (const ref of ["sk_test_CANARY", {kind:"secret_ref", instance_id:"foreign", integration:"clerk", output:"secret_key"}]) {
  it("rejects plaintext and foreign secret references", async () => {
    const client = new Client({run: () => ({status:0, stderr:"", stdout:JSON.stringify({ok:true, instance:"demo", instance_id:"owner-1", substrate:"local", integrations:{clerk:{secret_key:ref}}})})});
    await expect(client.up({kind:"create", on:"local"})).rejects.toThrow("invalid or foreign integration secret reference");
  });
}

it.each([[], "url", { public: { workload: "web", url: "https://example.test", source: "invented" } }, { public: { url: "https://example.test", source: "provider" } }])("rejects malformed endpoint bindings: %j", async (endpoints) => {
  const client = new Client({ bin: "/fake/stackless", run: () => ({ status: 0, stderr: "", stdout: JSON.stringify({ ok: true, instance: "demo", instance_id: "owner-1", substrate: "local", endpoints }) }) });
  await expect(client.up({ kind: "resume", name: "demo" })).rejects.toThrow("invalid endpoint binding");
});

it("binds endpoint URL maps through generated TypeScript names", async () => {
  const { bindEndpoints, BindError } = await import("../../../crates/stackless-idl/testdata/endpoints.js");
  const client = new Client({ bin: "/fake/stackless", run: () => ({ status: 0, stderr: "", stdout: JSON.stringify({
    ok: true, instance: "demo", instance_id: "owner-1", substrate: "local", endpoints: {
      "native-api": { workload: "web", url: "http://native.example.test", source: "provider" },
      "public-api": { workload: "web", url: "https://public.example.test/v1", source: "declared" },
    },
  }) }) });
  const outcome = await client.up({ kind: "resume", name: "demo" });
  expect(bindEndpoints(endpointUrls(outcome))).toEqual({ nativeApi: "http://native.example.test", publicApi: "https://public.example.test/v1" });
  expect(() => bindEndpoints({})).toThrow(BindError);
});

it.each([[], "local", {}, { workloads: { api: false }, resources: {} }, { workloads: {}, resources: { db: "" } }])("rejects malformed placements: %j", async (placements) => {
  const run: SpawnRunner = () => ({ status: 0, stderr: "", stdout: JSON.stringify({ ok: true, instance: "demo", instance_id: "owner-1", substrate: "local", origins: [], placements }) });
  await expect(new Client({ bin: "/fake/stackless", run }).up({ kind: "create", on: "local" })).rejects.toThrow(/invalid .*placement/);
});
